### PR TITLE
Add Backblaze B2 as remote provider

### PR DESCRIPTION
**Pull request by dpunkturban**
Hi, 

We use `pg_back` to back up and restore our databases to Backblaze. Currently we are using the S3 remote provider, which does not work very well with Backblaze, because sometimes backups randomly fail due to multipart upload failures. This is not a problem with pg_back, it is a problem with Backblaze (we find other tools on the internet that have exactly the same problem, such as https://github.com/wal-g/wal-g/issues/1741#issuecomment-2264672499). 

To avoid these problems and to speed up backups to Backblaze, I implemented the Backblaze native protocol B2 for `pg_back`, which is insanely faster than the S3 implementation and works out of the box. 

### Why Backblaze?
Backblaze is one of the largest blob storage backup providers, so I think it's very useful to support them. Other backup tools such as Restic also support the B2 protocol.

## Todos: 
- [x] Add Readme
- [x] Add config to `pg_back.conf`
- [x] Check naming consistency

### CODE REVIEW

**Comment by dpunkturban** [comment]
Hi @orgrim,

Is Backblaze B2 remote provider something you can imagine merging into the `pg_back` main? If yes, I would work on the open to-do points and then remove the PR draft.

Thanks :)

**Comment by orgrim** [comment]
Hi, I'm currently afk for some days, I'll review the PR as soon as I can.

Thanks for your patience

**Comment by dpunkturban** [comment]
> Hi, I'm currently afk for some days, I'll review the PR as soon as I can.
> 
> Thanks for your patience

Hi @orgrim, no prob, I will refactor the code in the next few days and work on the open todos. I will remove the draft when I'm finished :)

**Comment by orgrim** [comment]
Hi, so the tests run fine, I could run the docker tests locally. The docker tests are failing because github made a change on 
 actions in https://github.com/actions/runner-images/issues/9692

File: upload.go [hunk 1]
```diff
@@ -125,6 +144,118 @@ type s3repo struct {
...
+	localFile, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+
+	if _, err := io.Copy(file, remoteFile); err != nil {
+		localFile.Close()
+		return err
+	}
+	return localFile.Close()
```
Comment (orgrim) [inline upload.go:220]: Here, 2 local files are created and mixed, but only `file` gets data. It seems wrong.

Maybe remove the creation of `file` at the beginning of the function, and use `localFile` only, as the variable name is more readable.

File: upload.go [hunk 2]
```diff
@@ -125,6 +144,118 @@ type s3repo struct {
...
+func NewB2Repo(opts options) (*b2repo, error) {
+	r := &b2repo{
+		appKey:            opts.B2AppKey,
+		bucket:            opts.B2Bucket,
+		concurrentUploads: opts.B2ConcurrentUploads,
+		ctx:               context.Background(),
+		endpoint:          opts.B2Endpoint,
+		forcePath:         opts.B2ForcePath,
+		keyID:             opts.B2KeyID,
+		region:            opts.B2Region,
```
Comment (orgrim) [inline upload.go:156]: `region` is not used later in the code and the Backblaze Go API does not seem mention it, even if it seem to exist in the endpoint hostname they give on their WebUI

File: upload.go [hunk 3]
```diff
@@ -125,6 +144,118 @@ type s3repo struct {
 	session    *session.Session
 }
 
+func NewB2Repo(opts options) (*b2repo, error) {
+	r := &b2repo{
+		appKey:            opts.B2AppKey,
+		bucket:            opts.B2Bucket,
+		concurrentUploads: opts.B2ConcurrentUploads,
+		ctx:               context.Background(),
+		endpoint:          opts.B2Endpoint,
```
Comment (orgrim) [inline upload.go:153]: `endpoint` is not used but it seems possible to give a custom endpoint with `b2.APIBase()`

**Review by orgrim — CHANGES_REQUESTED** [review]
Great PR, the code works as expected, I could test it with a B2 bucket.

If the region is not used by B2, the `--b2-region` should be dropped. The `--b2-endpoint` is not used in the code by it seems good to allow to configure the endpoint

File: upload.go (hunk 1 above)
Comment (dpunkturban) [reply upload.go]: Hi,

fixed with including `defer X.close()`, i also did the same pattern in `Upload()`

File: upload.go (hunk 2 above)
Comment (dpunkturban) [reply upload.go]: Hi, fixed :)

File: upload.go (hunk 3 above)
Comment (dpunkturban) [reply upload.go]: Hi, I tried to find out if this is useful for anything and did not find why we need this. It looks like it is only necessary if we want to create S3 compatible URLs, where I see no use case for pg_back. I searched a little bit and all B2 compatible tools only use `keyID` and `appKey`. I think the region (US or EU) is baked into the keyID or bucket.

**Comment by dpunkturban** [comment]
Hey @orgrim, is there anything else needed to get this merged? thx

**Comment by orgrim** [comment]
> Hey @orgrim, is there anything else needed to get this merged? thx

Sorry for the delay, the PR is merged. Thanks a lot for this contribution!

**Comment by dpunkturban** [comment]
Hi @orgrim ,thanks for merging. Is it possible to create a new Release with the changes? :)

**Comment by orgrim** [comment]
Hi, yes but I need to have a look at the CI action that fails first, I hope to do that next week

**Comment by orgrim** [comment]
Hi @dpunkturban , release 2.5.0 is out!

**Pull request merged**